### ISSUE TITLE
Labels layer brush size not updating when layer switched with keybinding

### ISSUE THREAD

**Issue by ianhi** [napari/napari#2390]
## 🐛 Bug
When I switch between multiple labels layers using a keybinding the brush size as drawn on the canvas does not update correctly.
![Peek 2021-03-09 17-56](https://user-images.githubusercontent.com/10111092/110549911-c630c480-8100-11eb-9590-45a84a350005.gif)

## To Reproduce
1. Run this script:
```python
import napari
import numpy as np


with napari.gui_qt():
    arr = np.zeros([512, 512])
    arr1 = np.zeros([512, 512])
    arr2 = np.zeros([512, 512])

    # initialise viewer with astro image
    viewer = napari.view_image(arr)

    labels1 = viewer.add_labels(arr1, name='labels_1')
    labels2 = viewer.add_labels(arr2, name='labels_2')

    # Set the labels layer mode to picker with a string
    labels1.mode = 'paint'
    labels2.mode = 'paint'

    layer_arr = np.array([labels1, labels2])

    def set_correct_active_labels():
        """If a labels layer is active make sure it is the correct one"""
        new_layer = layer_arr[[labels1.visible, labels2.visible]][0]
        viewer.active_layer = new_layer

    def toggle(*args):
        labels1.visible = ~labels1.visible
        labels2.visible = ~labels2.visible
        set_correct_active_labels()

    viewer.bind_key("2", toggle)
```

2. Have a different brush size on the two different labels layers
3. Hit `2` to toggle between the layers.

## Expected behavior
Brush size updates when I change which labels layer is active.
## Environment
```
napari: 0.4.7.dev15+gdff1675.d20210308
Platform: Linux-5.8.0-7630-generic-x86_64-with-glibc2.31
System: Pop!_OS 20.04 LTS
Python: 3.9.1 | packaged by conda-forge | (default, Jan 26 2021, 01:34:10) [GCC 9.3.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.19.5
SciPy: 1.6.0
Dask: 2021.01.1
VisPy: 0.6.6

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 20.2.6
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- svg: 0.1.4
```

**Comment by sofroniewn** [comment]
I think this is related too https://github.com/napari/napari/issues/2356. The problem is with `viewer.active_layer = new_layer` it's really just not properly wired up. As mentioned in that issue we want to deprecate that PR, but in the mean time I guess we could fix, or just block the setter. Sorry about this. We can def try and address before `0.4.7`

**Comment by ianhi** [comment]
> or just block the setter. 

Ideally not this please! The partially working state is far better than not being able to switch it all. Losing the ability to bind this action to a keybind would make it super difficult to any sort of manual correction for situations with multiple layers (e.g. points + mask for watershed corrections).

It's also workaroundable, I've  attached a scroll callback to change the brush size which forces a rerender of the brush size.

> I think this is related too #2356. The problem is with viewer.active_layer = new_layer it's really just not properly wired up. 

Yeah there are a few other weird behaviors I've noticed but never isolated that I think are related to that as well. Especially when swapping visible labels layers and whether the active layer is a points or labels layer all at the same time.

**Comment by ianhi** [comment]
> It's also workaroundable, I've attached a scroll callback to change the brush size which forces a rerender of the brush size.

Here too I've noticed something that I wasn't sure should be raised as it's own issue or not. If you set brush size programmatically there doesn't seem to be an upper limit, so it can much larger than the napari window and cause wacky graphical errors - for example everything except napari blacking out for a few seconds.

Should I open issue for things like that? I feel like I've been opening so many and don't want to spam

**Comment by sofroniewn** [comment]
you can also manually unselect all the layers and select the one of interest so

```python
viewer.layers.unselect_all()
layer.selected = True
```

will have the desired effect of making `layer` now be the `viewer.active_layer`

**Comment by sofroniewn** [comment]
> If you set brush size programmatically there doesn't seem to be an upper limit, so it can much larger than the napari window and cause wacky graphical errors - for example everything except napari blacking out for a few seconds.

Yeah, that's not so great ...

> Should I open issue for things like that? I feel like I've been opening so many and don't want to spam

Yeah, I don't mind your issues are well written with example code/ screenshots etc which I appreciate. We won't be able to fix all of them immediately, but in 3 months someone should be able to come back to them and understand whats going on without have to dig through every last comment. Thanks for being aware of this though and asking, it is appreciated

**Comment by ianhi** [comment]
> you can also manually unselect all the layers and select the one of interest so

That's great! That also fixes the issue of brush size. I suppose this is equivalent to just clicking on the layer? 

Where should/can things like this be documented? I'm happy to write snippets of docs/examples for things like this but I'm not really sure where it should go.

**Comment by sofroniewn** [comment]
> That's great! That also fixes the issue of brush size. I suppose this is equivalent to just clicking on the layer?

Yes that's right.

> Where should/can things like this be documented? I'm happy to write snippets of docs/examples for things like this but I'm not really sure where it should go.

We're probably going to change all this in #2293, but it in general stuff like this could go in the viewer tutorial - see https://github.com/napari/tutorials/blob/master/fundamentals/viewer.ipynb. Contributions welcome there!

**Comment by ianhi** [comment]
> We won't be able to fix all of them immediately, but in 3 months someone should be able to come back to them and understand whats going on without have to dig through every last comment.

Honestly the level of responsiveness has been super gratifying. And I certainly have no expectation of most of them being fixed quickly or anything

**Comment by sofroniewn** [comment]
This is ultimately related to our `active_layer` issues which are being addressed in #2441. As there are other issues about that already, I will flag as a duplicate and close now
